packages/react-interactions/events/src/dom/create-event-handle/useEvent.js

```diff
@@ -30,20 +30,48 @@ export default function useEvent(
   |},
 ): UseEventHandle {
   const handleRef = useRef(null);
+  let setListener;
+  let clears;
+  let useEventHandle;
 
   if (handleRef.current == null) {
-    handleRef.current = createEventHandle(event, options);
+    setListener = createEventHandle(event, options);
+    clears = new Map();
+    useEventHandle = {
+      setListener(
+        target: EventTarget,
+        callback: null | ((SyntheticEvent<EventTarget>) => void),
+      ): void {
+        let clear = clears.get(target);
+        if (clear !== undefined) {
+          clear();
+        }
+        if (callback === null) {
+          clears.delete(target);
+          return;
+        }
+        clear = setListener(target, callback);
```

Comment:
Hmm, so you can only attach a single listener to a given target for a specific event type? I wonder if this constraint might cause any composability issues with custom hooks. It's a little confusing since non-`useEvent`-event handler uniqueness is the combination of target + type + function, where as this... adding a listener to a target would silently remove a previously added listener. Why?

Reply:
This is how `useEvent` worked before, calling `setListener` would mean updating the listener to the new callback you pass, removing the old one. Passing `null` would remove the listener altogther. The difference weith this change is that we first unmount the old listener, before mounting the new one. It's now expected that the user would put the callback for the event listener in their dependencies array (for their `useLayoutEffect` or `useEffect`, so this should only occur if the callback actually changes).

Reply:
The case I was concerned about was ... some part of code listening to an e.g. "click" handler and then another bit of code also listening and unwittingly removing the first handler. I guess you'll just need to be careful to avoid sharing `setListener` between code.

Reply:
Yeah, that's another reason why I wanted to change `createEventHandle` – so at least it follows the mental model better. You subscribe to event X and get back an unsubscribe function from it.

In terms of `useEvent`, I agree the mental model changes. It's less about `addEventListener` here, more about "write-less-code" when hooking up events, and have the lifecycles clean up for you.

Reply:
> Yeah, that's another reason why I wanted to change `createEventHandle` – so at least it follows the mental model better. You subscribe to event X and get back an unsubscribe function from it.

This make sense.

> In terms of `useEvent`, I agree the mental model changes. It's less about `addEventListener` here, more about "write-less-code" when hooking up events, and have the lifecycles clean up for you.

Maybe we could help here- avoid potential confusion- by requiring an explicit `clear` before `setListener` is called with a new function (and warning if you ever call `setListener` twice without clearing in between)? Seems like that would map nicely with the effects create+destroy pattern anyway.

Reply:
I think we should tackle changing `useEvent` in a seperate PR, especially as it involves changing lots of internals on www, which blocks the syncs unless the changes are merged together with the infra changes needed to the `useEvent` on www.

Reply:
Okay~ so long as we follow up on it? :smile:

Calling `setListener` twice without `clear` in between seems like an anti-pattern. It indicates you aren't using the effect cleanup callback correctly, or that you've written code that interferes with other code. I think we should have a warning for it.

Reply:
@bvaughn It depends what you're calling `setListener` on right?

You might be using the same `target` and wanting to update the listener to a new callback, in which case, that's fine.

Calling `clear` tears down all targets. Which means if you've used different targets then they also get affected. Example:

```jsx
function ConfirmationButtons({onAccept, onDecline}) {
  const clickHandle = useEvent('click');
  const acceptRef = useRef(null);
  const declineRef = useRef(null);

  useEffect(() => {
    clickHandle.setListener(acceptRef.current, onAccept);
    clickHandle.setListener(declineRef.current, onDecline);
  }, [onAccept, onDecline]);

  return (
    <>
      <button ref={declineRef}>Decline</button>
      <button ref={acceptRef}>Accept</button>
    </>
  )
}
```

The above pattern is meant to be promoted and you shouldn't need a `clear` anywhere here. `useEvent` automatically clears down its events when `ConfirmationButtons` gets unmounted.

Reply:
> The above pattern is meant to be promoted and you shouldn't need a clear anywhere here. useEvent automatically clears down its events when ConfirmationButtons gets unmounted.

I think I overlooked the fact that you were auto clearing. I think I read it as you just clearing the Map. (`clears.clear()` is used in a lot of places.)